instance/models/load_balancer.py

```diff
@@ -0,0 +1,130 @@
+
+    domain = models.CharField(max_length=100, unique=True)
+    # The username used to ssh into the server
+    ssh_username = models.CharField(max_length=32)
+    # Whether new backends can be assigned to this load-balancing server
+    accepts_new_backends = models.BooleanField(default=True)
+    # A random postfix appended to the haproxy configuration file names to make collisions
+    # impossible.
+    fragment_name_postfix = models.CharField(max_length=8, blank=True)
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.logger = ModelLoggerAdapter(logger, {'obj': self})
+
+    def format_log_message(self, msg):
+        """Format log messages for the load-balancing server."""
+        return "load_balancer={} ({!s:.15}) | {}".format(self.pk, self.domain, msg)
+
+    def get_configuration(self):
+        """Collect the backend maps and configuration fragments from all assoicated instances."""
+        backend_map = []
+        backend_conf = []
+        for instance in self.openedxinstance_set.iterator():
+            map_entries, conf = instance.get_load_balancer_configuration()
+            backend_map.append(map_entries)
+            backend_conf.append(conf)
+        return "\n\n".join(backend_map), "\n\n".join(backend_conf)
+
+    def get_fragment_name(self):
+        """Return the base name to use for the haproxy configuration files.
+
+        This should be unique for each instance of the instance manager, so that different
+        instance managers can share a load balancing server.
+        """
+        if not self.fragment_name_postfix:
+            bits = self._meta.get_field("fragment_name_postfix").max_length * 4
+            self.fragment_name_postfix = format(random.getrandbits(bits), "x")
+            self.save()
+        return settings.LOAD_BALANCER_FRAGMENT_NAME_PREFIX + self.fragment_name_postfix
+
+    def get_config_script(self):
+        """Render the configuration script to be executed on the loader balancer."""
+        backend_map, backend_conf = self.get_configuration()
+        fragment_name = self.get_fragment_name()
+        template = loader.get_template("instance/haproxy/conf.sh")
+        return template.render(dict(
+            settings=settings,
+            conf_filename=os.path.join(settings.LOAD_BALANCER_CONF_DIR, fragment_name),
+            backend_conf=backend_conf,
+            backend_filename=os.path.join(settings.LOAD_BALANCER_BACKENDS_DIR, fragment_name),
+            backend_map=backend_map,
+        ))
+
+    def reconfigure(self):
+        """Regenerate the configuration fragments on the load-balancing server."""
+        self.logger.info("Reconfiguring load-balancing server %s", self.domain)
+        config_script = self.get_config_script()
+        try:
+            subprocess.run(
+                ["ssh", "-T", "-o", "PasswordAuthentication=no", "-l", self.ssh_username, self.domain, "sudo sh"],
```

Comment:
Could you do this via ansible, like @bdero did in https://github.com/open-craft/opencraft/pull/144 ? It would be good to use a single approach for executing commands on a VM, and to factorize the common parts.